src/emqx_mgmt_cli.erl

```diff
@@ -550,9 +550,43 @@ listeners(["stop", Proto, ListenOn]) ->
             emqx_ctl:print("Failed to stop ~s listener on ~s, error:~p~n", [Proto, ListenOn, Error])
     end;
 
+listeners(["start",  Name = "http" ++ _N, ListenOn]) ->
+    case minirest:start_http(list_to_atom(Name)) of
+        ok ->
+            emqx_ctl:print("Start ~s listener on ~s successfully.~n", [Name, ListenOn]);
+        {error, Error} ->
+            emqx_ctl:print("Failed to start ~s listener on ~s, error:~p~n", [Name, ListenOn, Error])
+    end;
+
+listeners(["start", Proto, ListenOn]) ->
+    ListenOn1 = case string:tokens(ListenOn, ":") of
+        [Port]     -> list_to_integer(Port);
+        [IP, Port] -> {IP, list_to_integer(Port)}
+    end,
+    case emqx_listeners:start_listener({list_to_atom(Proto), ListenOn1, []}) of
```

Comment:
The `Options` parameter should not be an empty value. The value is derived from some content in the emqx application env. It cannot be empty because emqx_connection/emqx_ws_connection depends on a lot of it.

Reply:
Hi @HJianBo 
Sorry I'm a total erlang noob, please bear with me :)
I [searched](https://github.com/search?q=org%3Aemqx+start_listener&type=code) for usage of `start_listeners`, and the best I could find that matches your description is [`lists:foreach(fun start_listener/1, emqx:get_env(listeners, []))`](https://github.com/emqx/emqx/blob/7b98f62f7acbb000c899ebc64010e784894fc8c1/src/emqx_listeners.erl#L45).

As I said I'm completely new to erlang, so would need some help here. I would try to get all listeners from env `emqx:get_env(listeners, [])`, then something like
```
lists:filter(<find matching listener by proto+port>, emqx:get_env(listeners, []))
```
and then I think I could restart with the option found above. Am I on the right track?

Reply:
Yea, this is the only way to go in version 4.2 or 4.3

Reply:
@HJianBo best I could do is this
```
  listeners(["restart", Proto, ListenOn]) ->
      ListenOn1 = case string:tokens(ListenOn, ":") of
          [Port]     -> list_to_integer(Port);
          [IP, Port] -> {IP, list_to_integer(Port)}
      end,
      Listener = foldl(fun({LisProtocol, LisListenOn, LisOpts}, Acc) ->
          if
          (LisProtocol == Proto) andalso LisListenOn == ListenOn1 -> {LisProtocol, LisLis  tenOn, LisOpts};
          true -> Acc
          end
      end, {}, emqx:get_env(listeners, [])),
      case emqx_listeners:restart_listener(Listener) of
          ok ->
              emqx_ctl:print("Restarted ~s listener on ~s successfully.~n", [Proto, Liste  nOn]);
          {error, Error} ->
              emqx_ctl:print("Failed to restart ~s listener on ~s, error:~p~n", [Proto, L  istenOn, Error])
      end;
```
however it doesn't work, it never finds a listener from `emqx:get_env(listeners, [])`, I guess the comparison isn't right.

Reply:
Yes, the data type of `Proto` is a list, but `LisProtocol` is an atom. In addition to this, there are also some points that need attention:
1. the comparison of `ListenOn` is a bit more complicated. The user fills in the configuration file with 1883, but we actually change it to `{{0,0,0,0}, 1883}`. We need to be compatible with this change.
2. In addition to using `emqx_listeners` to restart listeners, we need to handle all listener output from `emqx_ctl listeners` that can be handled (even if it's simply reporting an unsupported error)

Reply:
@HJianBo ok I tried applying 1., with `list_to_atom` and making the `{{0,0,0,0}, list_to_integer(Port)}` tuple, but I still can't get to find the listener.

```
  listeners(["restart", Proto, ListenOn]) ->
      ListenOn1 = case string:tokens(ListenOn, ":") of
          [Port]     -> {{0,0,0,0}, list_to_integer(Port)};
          [IP, Port] -> {IP, list_to_integer(Port)}
      end,
      case find_listener_opts(Proto, ListenOn1) of
          {ok, Listener} ->
              case emqx_listeners:restart_listener(Listener) of
                  ok ->
                      emqx_ctl:print("Restarted ~s listener on ~s successfully.~n", [  Proto, ListenOn]);
                  {error, Error} ->
                      emqx_ctl:print("Failed to restart ~s listener on ~s, error:~p~n  ", [Proto, ListenOn, Error])
              end;
          {error, FindError} ->
              emqx_ctl:print("Failed to find ~s listener on ~s, error:~p~n", [Proto,   ListenOn, FindError])
      end;

  find_listener_opts(Proto, ListenOn) ->
      ProtoAtom = list_to_atom(Proto),
      foldl(fun({LisProtocol, LisListenOn, LisOpts}, Acc) ->
          if
          LisProtocol == ProtoAtom andalso LisListenOn == ListenOn ->
              {ok, {LisProtocol, LisListenOn, LisOpts}};
          true ->
              Acc
          end
      end, {error, "Listener/options not found"}, emqx:get_env(listeners, [])).
```

Re: 2., I really don't know how to go on about handling any more output outside of `ok` result...

Reply:
I still think it's `ListenOn1` that's not quite right. You run `bin/emqx attach` on an emqx that is already running to get to the emqx console. And execute `emqx:get_env(listeners, []). ` command to get to this data structure.

Reply:
@HJianBo you are right, thanks for the `emqx attach` hint.
```
      ListenOn1 = case string:tokens(ListenOn, ":") of
          [Port]     -> list_to_integer(Port);
          [IP, Port] -> {IP, list_to_integer(Port)}
      end,
```
I have been able to try it out and it works.

Do you think this is ready, or do we need anything else done here?